File: golang/q800/q852.go
package q800

func peakIndexInMountainArrayPlanA(arr []int) int {
	l := len(arr) - 1
	for i := 1; i < l; i++ {
		if arr[i-1] < arr[i] && arr[i] > arr[i+1] {
			return i
		}
	}
	return 0
}

func peakIndexInMountainArrayPlanB(arr []int) int {
	l, r := 0, len(arr)-1
	for l < r {
		m := (l + r) / 2
		if arr[m] < arr[m+1] {
			l = m + 1
		} else {
			r = m
		}
	}
	return l
}

func peakIndexInMountainArrayPlanC(arr []int) int {
	l, max, index := len(arr)-1, -1, -1
	for i := 1; i < l; i++ {
		if arr[i] > max {
			max = arr[i]
			index = i
		}
	}
	return index
}

File: hs/src/Q800/Q852.hs
module Q800.Q852 (
    peakIndexInMountainArrayPlanB,
    peakIndexInMountainArrayPlanC
) where

peakIndexInMountainArrayPlanB :: [Int] -> Int
peakIndexInMountainArrayPlanB arr = top 0 (length arr -1)
    where
        top l r = if l < r
                  then let
                          m = (l+r) `div` 2
                       in
                          if (arr!!m) > (arr!!(m+1)) then top l m
                          else top (m+1) r
                  else l

peakIndexInMountainArrayPlanC :: [Int] -> Int
peakIndexInMountainArrayPlanC = third . foldl f (-1, -1, 0)
    where
        f (index,m,i) x = if m > x then (index,m,i+1)
                    else (i, x, i+1)
        third (a,_,_) = a
